packages/augur-core/source/contracts/reporting/RepPriceOracle.sol

```diff
@@ -45,73 +67,82 @@ contract RepPriceOracle is IRepPriceOracle, Initializable {
-        uint256 _repAccumulationDelta = _repAmountAccumulated - _exchangeData.repAmountAccumulated;
-        // TODO: This should not be possible normally. When the real Uniswap contracts are in this condition can be removed.
-        if (_repAccumulationDelta == 0) {
+        // The first time we have actual data from the exchange we need to simply record the acumulator value and return the default since we have no accurate _relative_ delta to base the real price off of
+        if (_exchangeData.repPriceAccumulated == 0) {
+            _exchangeData.blockNumber = _blockNumber;
+            _exchangeData.blockTimestamp = _blockTimestamp;
+            _exchangeData.repPriceAccumulated = _repPriceCumulative;
             return _exchangeData;
         }
-        uint256 _deltaPrice = _cashAccumulationDelta * 1 ether / _repAccumulationDelta;
 
-        uint256 _weight = getWeight(_blockNumber - _exchangeData.blockNumber);
-        uint256 _price = (((1 ether - _weight) * _exchangeData.price) + (_weight * _deltaPrice)) / 1 ether;
+        uint256 _blocksElapsed = _blockNumber - _exchangeData.blockNumber;
 
-        _exchangeData.price = _price;
-        _exchangeData.blockNumber = _blockNumber;
-        _exchangeData.cashAmountAccumulated = _cashAmountAccumulated;
-        _exchangeData.repAmountAccumulated = _repAmountAccumulated;
+        uint256 _price = (_repPriceCumulative - _exchangeData.repPriceAccumulated) * 10**18 / _blocksElapsed / Q112;
+        require(_price > 0, "Price should not be 0");
 
-        return _exchangeData;
-    }
+        uint256 _secondsElapsed = _blockTimestamp - _exchangeData.blockTimestamp;
+        uint256 _priceAverage = _price;
 
-    function getWeight(uint256 _blockDelta) public view returns (uint256) {
-        uint256 _weight = _blockDelta * 1 ether / tau;
-        if (_weight > 1 ether) {
-            return 1 ether;
+        if (_secondsElapsed < period) {
+            _priceAverage = (_exchangeData.price * (period - _secondsElapsed) + _price * _secondsElapsed) / period;
         }
-        return _weight;
+
+        _exchangeData.blockNumber = _blockNumber;
+        _exchangeData.blockTimestamp = _blockTimestamp;
+        _exchangeData.repPriceAccumulated = _repPriceCumulative;
+        _exchangeData.price = _priceAverage;
+        return _exchangeData;
     }
 
     function initializeUniverse(IV2ReputationToken _reputationToken) private {
-        exchangeData[address(_reputationToken)].exchange = getOrCreateUniswapExchange(_reputationToken);
+        uint256 _blockNumber = block.number;
+        IUniswapV2 _exchange = getOrCreateUniswapExchange(_reputationToken);
+        exchangeData[address(_reputationToken)].exchange = _exchange;
+        if (_blockNumber != _exchange.blockNumberLast()) {
+            _exchange.sync();
+        }
         uint256 _initialPrice = getInitialPrice(_reputationToken);
         exchangeData[address(_reputationToken)].price = _initialPrice;
```

Comment:
Ignoring the massively terrible idea that this upgradable exchange presents for the moment... calling `initializeUniverse` will reset the price back to its default state which I don't think is appropriate behavior during an exchange upgrade.  In general, I don't think we should be re-initializing at all in that scenario but instead should be just changing the exchange address we have on file and assuming that the previous exchange was reasonably priced.

Reply:
I suppose we do need to reset the internal `repPriceAccumulated` value to the new exchange accumulator as well.

Reply:
Hmm, I just noticed that you actually reset the price after calling this in the upgrade scenario... so maybe the rest of this is fine to reset...